### PR TITLE
Allow queries to be invoked multiple times; fix `groupJoin`

### PR DESCRIPTION
**Pull request by hausdorff**
More details in the commit messages. The problem essentially is as follows.

Suppose I have the following.

```typescript
export async function* range(start: number, end?: number) {
    let i = start;
    while (true) {
        if (end !== undefined && i >= end) {
            return;
        }
        yield i++;
    }
}
```

Suppose we invoke this function `const xs = range(0, 10);` Now, `xs` will always return the same iterator. Once that iterator is exhausted, that’s it—every subsequent use (_e.g._, in a `for (const x of xs)`) will act as if there are 0 elements in the sequence.

We’d like to make a version of `xs` that can be “re-used” in this way. Right now, to get a “new” iterator, we’d have to keep `range` around and the arguments required to re-invoke it. We’d like instead for `xs` to simply return a fresh iterator any time we ask. In this PR we do something like the following to accomplish this:

```typescript
export function range(start: number, end?: number) {
    return wrapIterable(async function*() {
        let i = start;
        while (true) {
            if (end !== undefined && i >= end) {
                return;
            }
            yield i++;
        }
    });
}
```

### CODE REVIEW

**Comment by hausdorff** [comment]
I should note some of the alternative approaches possible here.

The above "wrapping" approach exists because our `Operator` type is defined as so:

```typescript
export type Operator<TSource, TResult> = (source: AsyncIterable<TSource>) => AsyncIterable<TResult>;
```

The main consequence of this is that we have to turn the `function*() {...}` that exists in these operators (as in `range` above) in a call to `wrapIterator`/`from`, which "transforms" the function into an `AsyncIterable`. This leads to `Operator` being (IMO) a "natural-feeling" type, mapping an `AsyncIterable` to an `AsyncIterable`. But an alternative approach would be to define `Operator` as:

```typescript
export type Operator<TSource, TResult> = (source: AsyncIterable<TSource>) => (function(): AsyncIterable<TResult>);
```

This would allow us to do the "wrapping" in the queryable base class, thus removing the need for all the calls to `wrapIterator`. I'd rather wrap the type on each call.

Notable also is that neither of these prevent users from doing something like the following:

```typescript
const k = from(mymap.keys());
```

`keys` returns an `Iterable` that will always return the same `Iterator`, which a number of users _will_ stumble upon when they see that a second invocation does not produce a useful result.

It is not clear at this time how to "solve" this problem—or, indeed, whether the JS community would expect it to be solved at all, given that so many core JS APIs behave in the same way.

File: nodejs/query/sources.ts [hunk 1]
```diff
@@ -12,41 +12,52 @@
...
-    }
-
-    if (isIterable(iter)) {
-        for (const t of iter) {
-            yield t;
-        }
-    } else {
-        for await (const t of iter) {
-            yield t;
+export function range(start: number, end?: number) {
```
Comment (justinvp) [inline nodejs/query/sources.ts]: Include the return type in the signature?

**Comment by hausdorff** [comment]
@lukehoban what do you think about @justinvp's proposal above?

**Comment by hausdorff** [comment]
We're a bit unsure about what the "right" API should be here, so we'll keep evaluating as we actually use the SDK to build stuff.

**Pull request merged**